Test failures on MariaDB performance_schema: OFF

I would report as I got 2 test failures on my local environment.

I tried it on current master latest branch (4a20d58a5a5b8096da3a220a3994853ca26c79c7).
Do you know what's wrong on my environment?

Thanks!

OS: Fedora 27
MariaDB 10.2
Ruby: 2.5.1

```
$ cat /etc/fedora-release 
Fedora release 27 (Twenty Seven)

$ rpm -q mariadb-devel mariadb-server
mariadb-devel-10.2.12-5.fc27.x86_64
mariadb-server-10.2.12-5.fc27.x86_64
```

Start MariaDB daemon.

```
$ sudo systemctl start mariadb
```

```
$ which mysql
/usr/bin/mysql

$ rpm -qf /usr/bin/mysql
mariadb-10.2.12-5.fc27.x86_64

$ mysql -u root -e 'CREATE DATABASE /*M!50701 IF NOT EXISTS */ test'

$ which ruby
/usr/local/ruby-2.5.1/bin/ruby

$ ruby -v
ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux]

$ which bundle
/usr/local/ruby-2.5.1/bin/bundle

$ bundle -v
Bundler version 1.16.1

$ bundle install --path vendor/bundle --without benchmarks development

$ bundle list
Gems included by the bundle:
  * ast (2.4.0)
  * bundler (1.16.1)
  * diff-lcs (1.3)
  * eventmachine (1.2.5)
  * mysql2 (0.5.1)
  * parallel (1.12.1)
  * parser (2.5.1.0)
  * powerpack (0.1.1)
  * rainbow (2.2.2)
  * rake (10.4.2)
  * rake-compiler (1.0.4)
  * rspec (3.7.0)
  * rspec-core (3.7.1)
  * rspec-expectations (3.7.0)
  * rspec-mocks (3.7.0)
  * rspec-support (3.7.1)
  * rubocop (0.50.0)
  * ruby-progressbar (1.9.0)
  * unicode-display_width (1.3.2)
```

Run unit tests.

```
$ bundle exec rake spec
mkdir -p tmp/x86_64-linux/mysql2/2.5.1
cd tmp/x86_64-linux/mysql2/2.5.1
/usr/local/ruby-2.5.1/bin/ruby -I. ../../../../ext/mysql2/extconf.rb
checking for rb_absint_size()... yes
checking for rb_absint_singlebit_p()... yes
checking for rb_wait_for_single_fd()... yes
-----
Using mysql_config at /usr/bin/mysql_config
-----
checking for mysql.h... yes
checking for errmsg.h... yes
checking for SSL_MODE_DISABLED in mysql.h... no
checking for MYSQL_OPT_SSL_ENFORCE in mysql.h... yes
checking for MYSQL.net.vio in mysql.h... no
checking for MYSQL.net.pvio in mysql.h... yes
checking for MYSQL_ENABLE_CLEARTEXT_PLUGIN in mysql.h... yes
checking for SERVER_QUERY_NO_GOOD_INDEX_USED in mysql.h... yes
checking for SERVER_QUERY_NO_INDEX_USED in mysql.h... yes
checking for SERVER_QUERY_WAS_SLOW in mysql.h... yes
checking for MYSQL_OPTION_MULTI_STATEMENTS_ON in mysql.h... yes
checking for MYSQL_OPTION_MULTI_STATEMENTS_OFF in mysql.h... yes
checking for my_bool in mysql.h... yes
-----
Setting libpath to /usr/lib64/
-----
creating Makefile
cd -
cd tmp/x86_64-linux/mysql2/2.5.1
/usr/bin/gmake
compiling ../../../../ext/mysql2/client.c
../../../../ext/mysql2/client.c: In function ‘rb_set_ssl_mode_option’:
../../../../ext/mysql2/client.c:140:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
../../../../ext/mysql2/client.c: At top level:
cc1: warning: unrecognized command line option ‘-Wno-used-but-marked-unused’
cc1: warning: unrecognized command line option ‘-Wno-static-in-inline’
cc1: warning: unrecognized command line option ‘-Wno-reserved-id-macro’
cc1: warning: unrecognized command line option ‘-Wno-missing-variable-declarations’
cc1: warning: unrecognized command line option ‘-Wno-documentation-unknown-command’
cc1: warning: unrecognized command line option ‘-Wno-disabled-macro-expansion’
cc1: warning: unrecognized command line option ‘-Wno-covered-switch-default’
cc1: warning: unrecognized command line option ‘-Wno-conditional-uninitialized’
cc1: warning: unrecognized command line option ‘-Wno-self-assign’
cc1: warning: unrecognized command line option ‘-Wno-constant-logical-operand’
cc1: warning: unrecognized command line option ‘-Wno-parentheses-equality’
compiling ../../../../ext/mysql2/infile.c
compiling ../../../../ext/mysql2/mysql2_ext.c
compiling ../../../../ext/mysql2/result.c
compiling ../../../../ext/mysql2/statement.c
linking shared-object mysql2/mysql2.so
cd -
mkdir -p tmp/x86_64-linux/stage/lib/mysql2
cp CHANGELOG.md tmp/x86_64-linux/stage/CHANGELOG.md
cp LICENSE tmp/x86_64-linux/stage/LICENSE
cp README.md tmp/x86_64-linux/stage/README.md
mkdir -p tmp/x86_64-linux/stage/examples
cp examples/eventmachine.rb tmp/x86_64-linux/stage/examples/eventmachine.rb
cp examples/threaded.rb tmp/x86_64-linux/stage/examples/threaded.rb
mkdir -p tmp/x86_64-linux/stage/ext/mysql2
cp ext/mysql2/client.c tmp/x86_64-linux/stage/ext/mysql2/client.c
cp ext/mysql2/client.h tmp/x86_64-linux/stage/ext/mysql2/client.h
cp ext/mysql2/extconf.rb tmp/x86_64-linux/stage/ext/mysql2/extconf.rb
cp ext/mysql2/infile.c tmp/x86_64-linux/stage/ext/mysql2/infile.c
cp ext/mysql2/infile.h tmp/x86_64-linux/stage/ext/mysql2/infile.h
cp ext/mysql2/mysql2_ext.c tmp/x86_64-linux/stage/ext/mysql2/mysql2_ext.c
cp ext/mysql2/mysql2_ext.h tmp/x86_64-linux/stage/ext/mysql2/mysql2_ext.h
cp ext/mysql2/mysql_enc_name_to_ruby.h tmp/x86_64-linux/stage/ext/mysql2/mysql_enc_name_to_ruby.h
cp ext/mysql2/mysql_enc_to_ruby.h tmp/x86_64-linux/stage/ext/mysql2/mysql_enc_to_ruby.h
cp ext/mysql2/result.c tmp/x86_64-linux/stage/ext/mysql2/result.c
cp ext/mysql2/result.h tmp/x86_64-linux/stage/ext/mysql2/result.h
cp ext/mysql2/statement.c tmp/x86_64-linux/stage/ext/mysql2/statement.c
cp ext/mysql2/statement.h tmp/x86_64-linux/stage/ext/mysql2/statement.h
cp ext/mysql2/wait_for_single_fd.h tmp/x86_64-linux/stage/ext/mysql2/wait_for_single_fd.h
cp lib/mysql2.rb tmp/x86_64-linux/stage/lib/mysql2.rb
cp lib/mysql2/client.rb tmp/x86_64-linux/stage/lib/mysql2/client.rb
cp lib/mysql2/console.rb tmp/x86_64-linux/stage/lib/mysql2/console.rb
cp lib/mysql2/em.rb tmp/x86_64-linux/stage/lib/mysql2/em.rb
cp lib/mysql2/error.rb tmp/x86_64-linux/stage/lib/mysql2/error.rb
cp lib/mysql2/field.rb tmp/x86_64-linux/stage/lib/mysql2/field.rb
cp lib/mysql2/result.rb tmp/x86_64-linux/stage/lib/mysql2/result.rb
cp lib/mysql2/statement.rb tmp/x86_64-linux/stage/lib/mysql2/statement.rb
cp lib/mysql2/version.rb tmp/x86_64-linux/stage/lib/mysql2/version.rb
mkdir -p tmp/x86_64-linux/stage/spec
cp spec/configuration.yml.example tmp/x86_64-linux/stage/spec/configuration.yml.example
mkdir -p tmp/x86_64-linux/stage/spec/em
cp spec/em/em_spec.rb tmp/x86_64-linux/stage/spec/em/em_spec.rb
cp spec/my.cnf.example tmp/x86_64-linux/stage/spec/my.cnf.example
mkdir -p tmp/x86_64-linux/stage/spec/mysql2
cp spec/mysql2/client_spec.rb tmp/x86_64-linux/stage/spec/mysql2/client_spec.rb
cp spec/mysql2/error_spec.rb tmp/x86_64-linux/stage/spec/mysql2/error_spec.rb
cp spec/mysql2/result_spec.rb tmp/x86_64-linux/stage/spec/mysql2/result_spec.rb
cp spec/mysql2/statement_spec.rb tmp/x86_64-linux/stage/spec/mysql2/statement_spec.rb
cp spec/rcov.opts tmp/x86_64-linux/stage/spec/rcov.opts
cp spec/spec_helper.rb tmp/x86_64-linux/stage/spec/spec_helper.rb
mkdir -p tmp/x86_64-linux/stage/spec/ssl
cp spec/ssl/ca-cert.pem tmp/x86_64-linux/stage/spec/ssl/ca-cert.pem
cp spec/ssl/ca-key.pem tmp/x86_64-linux/stage/spec/ssl/ca-key.pem
cp spec/ssl/ca.cnf tmp/x86_64-linux/stage/spec/ssl/ca.cnf
cp spec/ssl/cert.cnf tmp/x86_64-linux/stage/spec/ssl/cert.cnf
cp spec/ssl/client-cert.pem tmp/x86_64-linux/stage/spec/ssl/client-cert.pem
cp spec/ssl/client-key.pem tmp/x86_64-linux/stage/spec/ssl/client-key.pem
cp spec/ssl/client-req.pem tmp/x86_64-linux/stage/spec/ssl/client-req.pem
cp spec/ssl/gen_certs.sh tmp/x86_64-linux/stage/spec/ssl/gen_certs.sh
cp spec/ssl/pkcs8-client-key.pem tmp/x86_64-linux/stage/spec/ssl/pkcs8-client-key.pem
cp spec/ssl/pkcs8-server-key.pem tmp/x86_64-linux/stage/spec/ssl/pkcs8-server-key.pem
cp spec/ssl/server-cert.pem tmp/x86_64-linux/stage/spec/ssl/server-cert.pem
cp spec/ssl/server-key.pem tmp/x86_64-linux/stage/spec/ssl/server-key.pem
cp spec/ssl/server-req.pem tmp/x86_64-linux/stage/spec/ssl/server-req.pem
cp spec/test_data tmp/x86_64-linux/stage/spec/test_data
mkdir -p tmp/x86_64-linux/stage/support
cp support/5072E1F5.asc tmp/x86_64-linux/stage/support/5072E1F5.asc
cp support/libmysql.def tmp/x86_64-linux/stage/support/libmysql.def
cp support/mysql_enc_to_ruby.rb tmp/x86_64-linux/stage/support/mysql_enc_to_ruby.rb
cp support/ruby_enc_to_mysql.rb tmp/x86_64-linux/stage/support/ruby_enc_to_mysql.rb
install -c tmp/x86_64-linux/mysql2/2.5.1/mysql2.so lib/mysql2/mysql2.so
cp tmp/x86_64-linux/mysql2/2.5.1/mysql2.so tmp/x86_64-linux/stage/lib/mysql2/mysql2.so
/usr/local/ruby-2.5.1/bin/ruby -I/home/jaruga/git/mysql2/vendor/bundle/ruby/2.5.0/gems/rspec-core-3.7.1/lib:/home/jaruga/git/mysql2/vendor/bundle/ruby/2.5.0/gems/rspec-support-3.7.1/lib /home/jaruga/git/mysql2/vendor/bundle/ruby/2.5.0/gems/rspec-core-3.7.1/exe/rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb

Randomized with seed 54596
...
Failures:

  1) Mysql2::Client should set default program_name in connect_attrs
     Failure/Error: expect(result.first['attr_value']).to eq($PROGRAM_NAME)
     
     NoMethodError:
       undefined method `[]' for nil:NilClass
     # ./spec/mysql2/client_spec.rb:480:in `block (2 levels) in <top (required)>'

  2) Mysql2::Client should set custom connect_attrs
     Failure/Error: expect(results['program_name']).to eq('my_program_name')
     
       expected: "my_program_name"
            got: nil
     
       (compared using ==)
     # ./spec/mysql2/client_spec.rb:489:in `block (2 levels) in <top (required)>'

Finished in 10.35 seconds (files took 0.16545 seconds to load)
325 examples, 2 failures, 1 pending

Failed examples:

rspec ./spec/mysql2/client_spec.rb:474 # Mysql2::Client should set default program_name in connect_attrs
rspec ./spec/mysql2/client_spec.rb:483 # Mysql2::Client should set custom connect_attrs

Randomized with seed 54596
```

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -40,9 +40,8 @@ jobs:
           # db: A DB's brew package name in macOS case.
           #     Set a name "db: 'name@X.Y'" when using an old version.
           # MariaDB lastet version
-          # Allow failure due to the following test failures.
-          # https://github.com/brianmario/mysql2/issues/965
-          # https://github.com/brianmario/mysql2/issues/1152
+          # Allow failure due to the following test failures that rarely happens.
+          # https://github.com/brianmario/mysql2/issues/1194
           - {os: macos-latest, ruby: 2.4, db: mariadb, allow-failure: true}
           # MySQL latest version
           # Allow failure due to the issue #1165.

--- a/.github/workflows/container.yml
+++ b/.github/workflows/container.yml
@@ -14,13 +14,9 @@ jobs:
           # CentOS 7 system Ruby is the fixed version 2.0.0.
           - {distro: centos, image: 'centos:7', name_extra: 'ruby 2.0.0'}
           # Fedora latest stable version
-          # Allow failure due to the following test failures.
-          # https://github.com/brianmario/mysql2/issues/965
-          - {distro: fedora, image: 'fedora:latest', allow-failure: true}
+          - {distro: fedora, image: 'fedora:latest'}
           # Fedora development version
-          # Allow failure due to the following test failures.
-          # https://github.com/brianmario/mysql2/issues/1152
-          - {distro: fedora, image: 'fedora:rawhide', allow-failure: true}
+          - {distro: fedora, image: 'fedora:rawhide'}
       # On the fail-fast: true, it cancels all in-progress jobs
       # if any matrix job fails unlike Travis fast_finish.
       fail-fast: false

--- a/.travis.yml
+++ b/.travis.yml
@@ -41,3 +41,6 @@ matrix:
           - mysql2gem.example.com
   fast_finish: true
   allow_failures:
+    # Allow failure due to a package repository not found for now.
+    # https://travis-ci.org/github/brianmario/mysql2/jobs/767276558#L1255
+    - env: DB=mysql55

--- a/ci/Dockerfile_fedora
+++ b/ci/Dockerfile_fedora
@@ -20,6 +20,7 @@ RUN dnf -yq install \
   redhat-rpm-config \
   ruby-devel \
   rubygem-bigdecimal \
-  rubygem-bundler
+  rubygem-bundler \
+  rubygem-json
 
 CMD bash ci/container.sh

--- a/spec/mysql2/client_spec.rb
+++ b/spec/mysql2/client_spec.rb
@@ -484,7 +484,7 @@ RSpec.describe Mysql2::Client do # rubocop:disable Metrics/BlockLength
   end
 
   it "should set default program_name in connect_attrs" do
-    skip("DON'T WORRY, THIS TEST PASSES - but PERFORMNCE SCHEMA is not enabled in your MySQL daemon.") unless performance_schema_enabled
+    skip("DON'T WORRY, THIS TEST PASSES - but PERFORMANCE SCHEMA is not enabled in your MySQL daemon.") unless performance_schema_enabled
     client = new_client
     if Mysql2::Client::CONNECT_ATTRS.zero? || client.server_info[:version].match(/10.[01].\d+-MariaDB/)
       pending('Both client and server versions must be MySQL 5.6 or MariaDB 10.2 or later.')
@@ -494,7 +494,7 @@ RSpec.describe Mysql2::Client do # rubocop:disable Metrics/BlockLength
   end
 
   it "should set custom connect_attrs" do
-    skip("DON'T WORRY, THIS TEST PASSES - but PERFORMNCE SCHEMA is not enabled in your MySQL daemon.") unless performance_schema_enabled
+    skip("DON'T WORRY, THIS TEST PASSES - but PERFORMANCE SCHEMA is not enabled in your MySQL daemon.") unless performance_schema_enabled
     client = new_client(connect_attrs: { program_name: 'my_program_name', foo: 'fooval', bar: 'barval' })
     if Mysql2::Client::CONNECT_ATTRS.zero? || client.server_info[:version].match(/10.[01].\d+-MariaDB/)
       pending('Both client and server versions must be MySQL 5.6 or MariaDB 10.2 or later.')

--- a/spec/mysql2/client_spec.rb
+++ b/spec/mysql2/client_spec.rb
@@ -1,6 +1,11 @@
 require 'spec_helper'
 
 RSpec.describe Mysql2::Client do # rubocop:disable Metrics/BlockLength
+  let(:performance_schema_enabled) do
+    performance_schema = @client.query "SHOW VARIABLES LIKE 'performance_schema'"
+    performance_schema.any? { |x| x['Value'] == 'ON' }
+  end
+
   context "using defaults file" do
     let(:cnf_file) { File.expand_path('../../my.cnf', __FILE__) }
 
@@ -479,6 +484,7 @@ RSpec.describe Mysql2::Client do # rubocop:disable Metrics/BlockLength
   end
 
   it "should set default program_name in connect_attrs" do
+    skip("DON'T WORRY, THIS TEST PASSES - but PERFORMNCE SCHEMA is not enabled in your MySQL daemon.") unless performance_schema_enabled
     client = new_client
     if Mysql2::Client::CONNECT_ATTRS.zero? || client.server_info[:version].match(/10.[01].\d+-MariaDB/)
       pending('Both client and server versions must be MySQL 5.6 or MariaDB 10.2 or later.')
@@ -488,6 +494,7 @@ RSpec.describe Mysql2::Client do # rubocop:disable Metrics/BlockLength
   end
 
   it "should set custom connect_attrs" do
+    skip("DON'T WORRY, THIS TEST PASSES - but PERFORMNCE SCHEMA is not enabled in your MySQL daemon.") unless performance_schema_enabled
     client = new_client(connect_attrs: { program_name: 'my_program_name', foo: 'fooval', bar: 'barval' })
     if Mysql2::Client::CONNECT_ATTRS.zero? || client.server_info[:version].match(/10.[01].\d+-MariaDB/)
       pending('Both client and server versions must be MySQL 5.6 or MariaDB 10.2 or later.')

--- a/spec/mysql2/error_spec.rb
+++ b/spec/mysql2/error_spec.rb
@@ -39,6 +39,10 @@ RSpec.describe Mysql2::Error do
       end
     end
 
+    let(:server_info) do
+      @client.server_info
+    end
+
     before do
       # sanity check
       expect(valid_utf8.encoding).to eql(Encoding::UTF_8)
@@ -56,7 +60,15 @@ RSpec.describe Mysql2::Error do
         expect(bad_err.message.encoding).to eql(Encoding::UTF_8)
         expect(bad_err.message).to be_valid_encoding
 
-        expect(bad_err.message).to include("??}\u001F")
+        # MariaDB 10.5 returns a little different error message unlike MySQL
+        # and other old MariaDBs.
+        # https://jira.mariadb.org/browse/MDEV-25400
+        err_str = if server_info[:version].match(/MariaDB/) && server_info[:id] >= 100500
+          "??}\\001F"
+        else
+          "??}\u001F"
+        end
+        expect(bad_err.message).to include(err_str)
       end
     end
 

--- a/spec/mysql2/statement_spec.rb
+++ b/spec/mysql2/statement_spec.rb
@@ -5,12 +5,19 @@ RSpec.describe Mysql2::Statement do
     @client = new_client(encoding: "utf8")
   end
 
+  let(:performance_schema_enabled) do
+    performance_schema = @client.query "SHOW VARIABLES LIKE 'performance_schema'"
+    performance_schema.any? { |x| x['Value'] == 'ON' }
+  end
+
   def stmt_count
     # Use the performance schema in MySQL 5.7 and above
-    @client.query("SELECT COUNT(1) AS count FROM performance_schema.prepared_statements_instances").first['count'].to_i
-  rescue Mysql2::Error
-    # Fall back to the global prepapred statement counter
-    @client.query("SHOW STATUS LIKE 'Prepared_stmt_count'").first['Value'].to_i
+    if performance_schema_enabled
+      @client.query("SELECT COUNT(1) AS count FROM performance_schema.prepared_statements_instances").first['count'].to_i
+    else
+      # Fall back to the global prepapred statement counter
+      @client.query("SHOW STATUS LIKE 'Prepared_stmt_count'").first['Value'].to_i
+    end
   end
 
   it "should create a statement" do

